heat/core/indexing.py

```diff
@@ -56,39 +56,45 @@ def nonzero(x: DNDarray) -> DNDarray:
     except AttributeError:
         raise TypeError("Input must be a DNDarray, is {}".format(type(x)))
 
+    lcl_nonzero = torch.nonzero(input=local_x, as_tuple=False)
+
     if x.split is None:
-        # if there is no split then just return the values from torch
-        lcl_nonzero = torch.nonzero(input=local_x, as_tuple=False)
+        # if there is no split then just return the transpose of values from torch
+        lcl_nonzero = lcl_nonzero.transpose(0, 1)
+
         gout = list(lcl_nonzero.size())
         is_split = None
     else:
         # a is split
-        lcl_nonzero = torch.nonzero(input=local_x, as_tuple=False)
         # adjust local indices along split dimension
         _, displs = x.counts_displs()
         lcl_nonzero[..., x.split] += displs[x.comm.rank]
         del displs
+
+        lcl_nonzero = lcl_nonzero.transpose(0, 1)
+
         # get global size of split dimension
         gout = list(lcl_nonzero.size())
         gout[0] = x.comm.allreduce(gout[0], MPI.SUM)
```

Comment:
I had written a lengthy detailed explanation of what's happening here, and then I lost it. Doh. Anyway.

The original `torch.nonzero` output shape is _(number_of_nonzero_elements,  number_of_dimensions)_. The process-local `lcl_nonzero` only know about how many nonzero elements are on process. But the global output DNDarray must know the total.

That `allreduce` call replaces the local value of `gout[0]` (local number of rows = local number of nonzero elements) with the sum of all `gout[0]`s on all processes, in order to synchronize `gout` to the global (albeit memory-distributed) number of rows of the output.  

Problem: `lcl_nonzero` has been transposed, so it's no longer the rows, but the columns representing the number of nonzero elements.

My suggestion: 
- transpose before the `if split` check
- adjust line 71 `lcl_nonzero[..., x.split] += displs[x.comm.rank]` to correct the row (not the column) corresponding to the split dimension
- the MPI collective call at line 78 should sum along the dimension containing the number of nonzero elements, now the columns